File: examples/canonical-terminal/ruby-clamp-tty/lib/taskflow/core/loader.rb
# frozen_string_literal: true

require "json"
require "pathname"
require_relative "job"

module Taskflow
  module Core
    class FixtureError < StandardError; end

    module Loader
      module_function

      def default_fixtures_path
        env_path = ENV["TASKFLOW_FIXTURES_PATH"]
        return Pathname(env_path).expand_path if env_path && !env_path.empty?

        Pathname(__dir__).join("../../../../fixtures").expand_path
      end

      def load_jobs(path = default_fixtures_path)
        load_json(path, "jobs.json").map { |item| Job.from_h(item) }
      end

      def load_config(path = default_fixtures_path)
        load_json(path, "config.json")
      end

      def load_json(path, filename)
        fixtures = Pathname(path).expand_path
        raise FixtureError, "fixtures path does not exist: #{fixtures}" unless fixtures.exist?

        target = fixtures.join(filename)
        raise FixtureError, "missing fixture file: #{target}" unless target.exist?

        JSON.parse(target.read)
      rescue JSON::ParserError => e
        raise FixtureError, "invalid json in #{target}: #{e.message}"
      end
    end
  end
end

File: examples/canonical-terminal/elixir-optimus-ratatouille/lib/taskflow/core/loader.ex
defmodule Taskflow.Core.Loader do
  alias Taskflow.Core.{Event, Job}

  defmodule FixtureError do
    defexception [:message]
  end

  def default_fixtures_path do
    System.get_env("TASKFLOW_FIXTURES_PATH") ||
      Path.expand("../../../../fixtures", __DIR__)
  end

  def load_jobs(path \\ default_fixtures_path()) do
    load_json(path, "jobs.json", &Job.from_map/1)
  end

  def load_events(path \\ default_fixtures_path()) do
    load_json(path, "events.json", &Event.from_map/1)
    |> Enum.sort_by(& &1.timestamp, DateTime)
  end

  def load_config(path \\ default_fixtures_path()) do
    load_json(path, "config.json", & &1)
  end

  defp load_json(path, filename, mapper) do
    fixtures_path = Path.expand(path)

    unless File.dir?(fixtures_path) do
      raise FixtureError, "fixtures path does not exist: #{fixtures_path}"
    end

    target = Path.join(fixtures_path, filename)

    unless File.exists?(target) do
      raise FixtureError, "missing fixture file: #{target}"
    end

    decode_json(target, mapper)
  end

  defp decode_json(target, mapper) do
    target
    |> File.read!()
    |> Jason.decode!()
    |> then(fn
      items when is_list(items) -> Enum.map(items, mapper)
      item -> mapper.(item)
    end)
  rescue
    e in Jason.DecodeError -> raise FixtureError, "invalid json in #{target}: #{Exception.message(e)}"
  end
end
